pkg/storage/redis/redis.go

```diff
@@ -208,6 +212,78 @@ func (db *DB) ClearDeleted(_ context.Context, cutoff time.Time) {
 	}
 }
 
+func doNotify(ctx context.Context, psc *redis.PubSubConn, ch chan struct{}) error {
+	switch v := psc.ReceiveWithTimeout(time.Second).(type) {
+	case redis.Message:
+		log.Debug().Str("action", string(v.Data)).Msg("Got redis message")
+		if string(v.Data) != watchAction {
+			return nil
+		}
+		select {
+		case <-ctx.Done():
+			log.Warn().Err(ctx.Err()).Msg("unable to notify channel")
+			return ctx.Err()
+		case ch <- struct{}{}:
+		}
+	case error:
+		log.Debug().Err(v).Msg("redis subscribe error")
+		return v
+	}
+	return nil
+}
+
+func (db *DB) watchLoop(ctx context.Context, ch chan struct{}, eb *backoff.ExponentialBackOff) {
+top:
+	psConn := db.pool.Get()
+	defer psConn.Close()
+	psc := redis.PubSubConn{Conn: psConn}
+	if err := psc.PSubscribe("__keyspace*__:" + db.versionSet); err != nil {
+		log.Error().Err(err).Msg("failed to subscribe to version set channel")
+		return
+	}
+
+	for {
+		err, ok := doNotify(ctx, &psc, ch).(net.Error)
+		if !ok && err != nil {
+			log.Error().Err(ctx.Err()).Msg("failed to notify channel")
+			return
+		}
+		if ok && err.Timeout() {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(eb.NextBackOff()):
+			}
+			goto top
```

Comment:
Can this be rewritten without using a `goto`?

Reply:
I'm not sure there's a cleaner way. In case of timeout, the underlying connection was closed, so we need to start from scratch, with a new fresh pub sub conn. This way looks cleaner for me, instead of using nested for loop, with a break here instead of goto.

Reply:
My concern (probably mostly from other languages) is that I have a hard time mentally modeling behavior/state with gotos. I know there is a time and place for them, but ... yeah, I'm not sure here.

@calebdoxsey thoughts?

Reply:
I don't think its a good idea to use goto like this.

Reply:
@calebdoxsey why and any suggestion to improve?

Reply:
> instead of using nested for loop, with a break here

Can we try this? 

Perhaps abstracting the flow into a few separate functions can help with the janky-ness of starting "from scratch, with a new fresh pub sub conn".

Reply:
@desimone fixed